src/cpp/src/whisper/whisper.cpp

```diff
@@ -0,0 +1,201 @@
+
+    request.set_tensor("input_features", input_tensor);
+
+    request.infer();
+
+    // reset input tensor
+    request.set_tensor("input_features", ov::Tensor(ov::element::f32, input_shape));
+
+    return request.get_tensor("last_hidden_state");
+}
+
+void set_past_kev_value(ov::InferRequest& source, ov::InferRequest& dest) {
+    // source outputs:
+    // present.0.decoder.key
+    // present.0.decoder.value
+    // present.0.encoder.key
+    // present.0.encoder.value
+
+    // dest imputs:
+    // past_key_values.0.decoder.key
+    // past_key_values.0.decoder.value
+    // past_key_values.0.encoder.key
+    // past_key_values.0.encoder.value
+
+    for (auto& dec_output : source.get_compiled_model().outputs()) {
+        std::string dec_output_name = dec_output.get_any_name();
+        if (dec_output_name.find("logits") != std::string::npos) {
+            continue;
+        }
+
+        std::string with_past_input_name =
+            std::regex_replace(dec_output_name, std::regex("present"), "past_key_values");
+
+        auto kv_tensor = source.get_tensor(dec_output_name);
+        dest.set_tensor(with_past_input_name, ov::Tensor{kv_tensor});
+    }
+}
+
+int64_t decode(ov::Tensor& encoder_hidden_state,
+               ov::InferRequest& decoder,
+               std::vector<int64_t>& input_ids,
+               const ov::genai::WhisperGenerationConfig& config) {
+    decoder.set_tensor("encoder_hidden_states", ov::Tensor{encoder_hidden_state});
+
+    ov::Tensor input_ids_tensor(ov::element::i64, {1, input_ids.size()}, input_ids.data());
+    decoder.set_tensor("input_ids", input_ids_tensor);
+
+    decoder.infer();
+
+    auto output_tensor = decoder.get_tensor("logits");
+
+    suppress_tokens(output_tensor, config.begin_suppress_tokens);
+
+    int64_t output_token = ov::genai::utils::argmax(output_tensor, 0);
+
+    return output_token;
+}
+
+int64_t decode_with_past(ov::Tensor& encoder_hidden_state,
+                         ov::InferRequest& decoder_with_past,
+                         int64_t input_id,
+                         const size_t cache_position,
+                         const ov::genai::WhisperGenerationConfig& config) {
+    decoder_with_past.set_tensor("encoder_hidden_states", ov::Tensor{encoder_hidden_state});
+
+    std::vector<int64_t> input_ids = {input_id};
+    ov::Tensor input_ids_tensor(ov::element::i64, {1, 1}, input_ids.data());
+    decoder_with_past.set_tensor("input_ids", input_ids_tensor);
+
+    ov::Tensor cache_position_tensor = decoder_with_past.get_tensor("cache_position");
+    cache_position_tensor.set_shape({1});
+    cache_position_tensor.data<int64_t>()[0] = cache_position;
+
+    decoder_with_past.infer();
+
+    auto output_tensor = decoder_with_past.get_tensor("logits");
+
+    suppress_tokens(output_tensor, config.suppress_tokens);
+
+    int64_t output_token = ov::genai::utils::argmax(output_tensor, 0);
+
+    set_past_kev_value(decoder_with_past, decoder_with_past);
```

Comment:
```suggestion
    set_past_key_value(decoder_with_past, decoder_with_past);
```
do we need to reset input / output on each iteration? 
I suppose it's allowed only once to "connect" them

Reply:
You're right, reimplemented to 'connect' them only once. Thanks!